libs/wingc/src/type_check.rs

```diff
@@ -398,6 +405,12 @@ impl Subtype for Type {
 				let r: &Type = &*r0;
 				l.is_subtype_of(r)
 			}
+			(Self::MutSet(l0), Self::Set(r0)) => {
+				// A MutSet type is a subtype of a Set type if the value type is a subtype of the other value type
+				let l: &Type = &*l0;
+				let r: &Type = &*r0;
+				l.is_subtype_of(r)
+			}
```

Comment:
That's a very good counter-example. 
Perhaps `MutArray` and `Array` should be siblings - inheriting from a common `BaseArray`. It's a little bit tricky when considering methods like `.concat()` which accept and return an array (and ideally we'd want to allow accepting both `MutArray` and `Array` as arguments).
I'll work on that in a subsequent PR.

For now, I'll remove these changes since the rest of the PR is a necessary refactoring to address std lib objects as method arguments and I'd like to add that unrelated to the good point you made.

Reply:
Mmmm interesting point. I think you are right that we should break that relationship (also in the SDK declarations).

I think that they should just be two separate types. No need for a base type I think.

Reply:
A common parent for `Array` and `MutArray` is desired to avoid code duplication of all the array api methods (e.g. `index_of`, `concat`, etc.).

Speaking of `.concat()`, it complicates things a bit because ideally, we would like to be able to accept both `Array` and `MutArray` as the argument:
```ts
let immut_arr1 = Array<num> [1,2,3];
let immut_arr2 = Array<num> [4,5,6];
let mut_arr = MutArray<num> [7,8,9];

// this is supported
immut_arr1.concat(immut_arr2);

// ideally, we'd like to support that as well 
immut_arr1.concat(mut_arr);

// the alternative is converting MutArray to Array
immut_arr1.concat(mut_arr.copy());
```

Reply:
Let's discuss what's the best approach. Sometimes a bit of duplication is the right thing. Especially when the code is a set of simple macros.

Reply:
We should have use the "is Foo a Bar" rule. We we had we would have noticed that the answer to "is an ImmutableArray a MutableArray?" is false and scrapped that implementation.
I wonder, can we avoid code duplication with an abstract class or generic in type script and then export the the concrete classes through JSII?

Reply:
I spoke with @eladb offline about this and went with breaking Array/MutArray, Set/MutSet, Map/MutMap to independent types. Added to this PR.